src/util.rs

```diff
@@ -55,3 +55,323 @@ impl Row for Index {
+            target -= cur & (diff >> 8);
+            // Descend to upper half or lower half
+            // depending on are we over or under
+            cur = (child >> (result - W(child_stride)).0 as usize) & W(child_mask);
+        };
+        descend(c, 4/*16*/,  8, 0b00001111);// PAR[3]
+        descend(b, 5/* 8*/,  4, 0b00000111);// PAR[2]
+        descend(a, 6/* 4*/,  2, 0b00000011);// PAR[1]
+        descend(n, 7/* 2*/,  1, 0b00000001);// PAR[0]
+    }
+    result -= (cur - target & W(256)) >> 8;
+
+    Some(result.0)
+}
+
+#[test]
+fn parity_0_average_ones_u32() {
+    struct EvenParity(u32);
+
+    impl Iterator for EvenParity {
+        type Item = u32;
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.0 == u32::max_value() {
+                return None;
+            }
+            self.0 += 1;
+            while self.0.count_ones() & 1 != 0 {
+                if self.0 == u32::max_value() {
+                    return None;
+                }
+                self.0 += 1;
+            }
+            Some(self.0)
+        }
+    }
+
+    let steps = 1000;
+    for i in 0..steps {
+        let pos = i * (u32::max_value() / steps);
+        for i in EvenParity(pos).take(steps as usize) {
+            let mask = (1 << (average_ones_u32(i).unwrap_or(32) - 1)) - 1;
+            assert_eq!(
+                (i & mask).count_ones(),
+                (i & !mask).count_ones(),
+                "{}", i
+            );
+        }
+    }
+}
+
+#[test]
+fn parity_1_average_ones_u32() {
+    struct OddParity(u32);
+
+    impl Iterator for OddParity {
+        type Item = u32;
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.0 == u32::max_value() {
+                return None;
+            }
+            self.0 += 1;
+            while self.0.count_ones() & 1 == 0 {
+                if self.0 == u32::max_value() {
+                    return None;
+                }
+                self.0 += 1;
+            }
+            Some(self.0)
+        }
+    }
+
+    let steps = 1000;
+    for i in 0..steps {
+        let pos = i * (u32::max_value() / steps);
+        for i in OddParity(pos).take(steps as usize) {
+            let mask = (1 << (average_ones_u32(i).unwrap_or(32) - 1)) - 1;
+            let a = (i & mask).count_ones();
+            let b = (i & !mask).count_ones();
+            if a < b {
+                assert_eq!(a + 1, b, "{:x}", i);
+            } else if b < a{
+                assert_eq!(a, b + 1, "{:x}", i);
+            } else {
+                panic!("Odd parity shouldn't split in exactly half");
+            }
+        }
+    }
+}
+
+#[test]
+fn empty_average_ones_u32() {
+    assert_eq!(None, average_ones_u32(0));
+}
+
+#[test]
+fn singleton_average_ones_u32() {
+    for i in 0..32 {
+        assert_eq!(
+            None,
+            average_ones_u32(1 << i),
+            "{:x}", i
+        );
+    }
+}
+
+#[allow(dead_code)]
+fn average_ones_u64(n: u64) -> Option<u64> {
```

Comment:
This is hard for me to understand

Reply:
It basically does parallel [bit count](http://bits.stephan-brumme.com/countBits.html) and then finding place where the bit count goes over half using branchless binary search.

Reply:
After some benchmarking, it seems that the parallel bit counting is slower than std's `count_ones`. Additionally, a very dumb search is faster than the branchless binary search. So in total we can cut the time in half.

Reply:
Hold on, for larger numbers it's the exact opposite.

Reply:
Interesting... Dunno if the speed is worth it for hard to understandness.
Splitting is not even that performance critical..

Reply:
I can also do branchful binary search.